extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/BigQueryServiceImpl.java

```diff
@@ -0,0 +1,594 @@
+        if (streamWriter == null) {
+            monitor.debug("BigQuery Sink creating stream writer in pending mode for table " + target.getTableName().toString());
+            var stream = WriteStream.newBuilder().setType(WriteStream.Type.PENDING).build();
+
+            var createWriteStreamRequest =
+                    CreateWriteStreamRequest.newBuilder()
+                    .setParent(target.getTableName().toString())
+                    .setWriteStream(stream)
+                    .build();
+            var writeStream = writeClient.createWriteStream(createWriteStreamRequest);
+
+            var builder = JsonStreamWriter.newBuilder(writeStream.getName(), writeClient);
+            if (credentials != null) {
+                builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+            }
+
+            streamWriter = builder.build();
+        }
+
+        if (streamWriter.isClosed()) {
+            // TODO re-create the writer for a maximum, specified number of times.
+            monitor.info("BigQuery Sink stream writer closed, recreating it for stream " + streamWriter.getStreamName());
+            var builder = JsonStreamWriter.newBuilder(streamWriter.getStreamName(), writeClient);
+            if (credentials != null) {
+                builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+            }
+            streamWriter = builder.build();
+        }
+    }
+
+    private void initService() throws IOException {
+        if (bigQuery != null && writeClient != null) {
+            return;
+        }
+
+        initCredentials();
+
+        var project = target.project();
+        if (project == null) {
+            project = gcpConfiguration.getProjectId();
+        }
+
+        var bqBuilder = BigQueryOptions.newBuilder().setProjectId(project);
+
+        if (credentials != null) {
+            bqBuilder.setCredentials(credentials);
+        }
+
+        bigQuery = bqBuilder.build().getService();
+
+        var settingsBuilder = BigQueryWriteSettings.newBuilder();
+        if (credentials != null) {
+            settingsBuilder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+        }
+
+        writeClient = BigQueryWriteClient.create(settingsBuilder.build());
+    }
+
+    private JSONObject buildRecord(FieldList fields, FieldValueList values) {
+        var record = new JSONObject();
+        int colCount = fields.size();
+        for (int i = 0; i < colCount; i++) {
+            record.put(fields.get(i).getName(), values.get(i).getValue());
+        }
+
+        return record;
+    }
+
+    private void refreshBigQueryCredentials() {
+        var credentials = bigQuery.getOptions().getCredentials();
+        if (!(credentials instanceof OAuth2Credentials)) {
+            return;
+        }
+
+        try {
+            // TODO check margin and refresh if too close to expire.
+            ((OAuth2Credentials) credentials).refreshIfExpired();
```

Comment:
if you flip the condition you could use the pattern variable statement and avoid the cast at line 195